### ISSUE TITLE
Request: alternative hashing algorithms for better performance

### ISSUE THREAD

**Issue by xkortex** [git-lfs/git-lfs#4461]
**Describe the issue**
The checksum stage of LFS clean/smudge is a major bottleneck (outset network transfer of course), and with multi-gigabytes of files tracked by LFS, this can lead to prohibitive times to perform some operations, such as branch changes, cleaning the repo to free space, and verifying LFS content. Faster hash algorithms have been developed. This is probably one of those "long-haul requests" but I figure it would be good to raise awareness now for exactly that reason. The second-best time to plant a tree is today. 

The current scheme is `sha256`, as evidenced by LFS metadata:

```
version https://git-lfs.github.com/spec/v1
oid sha256:a73295ac66f9cb74d558ce3ade46f75e2f2997ed05eeed0f4b774623372058ea
size 282442597
``` 

[BLAKE3](https://github.com/BLAKE3-team/BLAKE3) is a very modern hashing function (family of functions in fact). It routinely performs 10x-20x faster than `sha256`. If BLAKE3 could be implemented (and I'm making no claims of how easy or difficult that might be), this would result in _dramatically_ faster operations.

**Barriers to implementation**
- Checksum function would have to be parameterized. `hash.Hash` interface is already used which is great, though there's a handful of places where `sha256.New()` is called by name. I've only lightly dug through the repo and it looks like  `oidType` is flexible, so the infrastructure to support multiple hash types might largely already exist (good thinking ahead!) 
- Availability of BLAKE3/b3sum. Since it's relatively new, it isn't included in x/crypto at the moment. However there is a [feature request](https://github.com/golang/go/issues/36632). It's probably going to pull in https://github.com/zeebo/blake3 or https://github.com/lukechampine/blake3
- The biggest hurdle by far would be gracefully interoperating with different hashing schemes. This would depend on how abstracted the oidParser is, looks like right now it's just checking against the hardcoded value: 

https://github.com/git-lfs/git-lfs/blob/main/lfs/pointer.go#L195

But it seems relatively tractable. `parseOid` would just need to check against a slice of valid prefixes, and pass `oidType` along with the `oid` (either would need something like `NewGeneralPointer`, or change the interface of `NewPointer`, whichever is more palatable). One may want to also check that `OidType` matches in addition to `Oid`, although I would imagine the collision probability between sha256 and blake3 hashes is diminutive.

**Comment by bk2204** [comment]
Hey,

So, first of all, we generally try to minimize the amount of hashing done.  Usually hashing is not done unless a file is being cleaned or transferred.  In the latter case, the amount of time spent hashing is almost certainly dwarfed by the amount of time being spent on transfer except in the fastest networks.  It is true that in your particular case, where you rsync working trees across the network, that does result in a variety of files being cleaned at great expense, but [the Git FAQ is in the process of being updated](https://lore.kernel.org/git/20210227191813.96148-2-sandals@crustytoothpaste.net/) to mention that being expensive and therefore not recommended.  Note that this would also be the case if you were using raw large objects in Git, but worse, since Go's SHA-256 implementation is almost certainly faster than Git's SHA-1-DC implementation which checks for SHA-1 collisions.  It is not specific to Git LFS in any way.

So in that context, I'd want to see some benchmarks about where time is spent before we add additional hash algorithms.  I'd also like to point out that on newer AMD and ARM (and some Intel) machines, SHA-256 is accelerated in hardware, and is therefore likely to be the fastest option.  It is true that many Intel processors lack those instructions now, but their existence is part of the reason that Git used SHA-256 as its new hash algorithm to replace SHA-1.

If we decide we want to support an additional hash algorithm, I think I'd prefer to go with BLAKE2b.  BLAKE3 has been weakened a bit too much for me to have confidence in its long-term viability and therefore I'm uncomfortable with shipping software based on it.  I think BLAKE2b is commonly used enough that it's still an interesting target for cryptanalysis and has a good tradeoff between security and speed.

Even if we ship support for a new algorithm, that doesn't mean that you'll be able to use it everywhere.  For example, I happen to know that some parts of the GitHub LFS implementation store objects directly by their SHA-256 hash.  Adding support for literally any other algorithm will require major changes to how that code works and stores objects, probably necessitating a major redesign.  Implementations that use Amazon S3 will also generally want to use SHA-256 because it makes uploading objects via signed URLs much easier.  So there isn't a guarantee that if you create such a repository that it will be possible to use it everywhere, or that you'll be able to upload your LFS files to an arbitrary location, at least at first.

I should also point out that in many contexts the use of certain hash algorithms is requires (e.g., by government or industry standards).  So even if we shipped BLAKE2b support, that doesn't mean that it would be universally used or that it could be the default.  We also would not mix algorithms in a repository, so you'd only get the benefits for new repositories.

So this is definitely a possibility for us to implement, but I do want to be clear that it is not likely to be implemented anytime soon.  I'll mark this accordingly as an enhancement.

**Comment by xkortex** [comment]
That's about what I anticipated, I am certainly not holding out for this feature. 

I didn't know about the hardware SHA256 support - if that's the case, then probably any advantage of software hash algorithms is going to be wiped out entirely. 

I'll try to get some profiling, but I don't think I can currently test on anything with that hardware acceleration.  

> I should also point out that in many contexts the use of certain hash algorithms is requires (e.g., by government or industry standards).

That would seem to suggest parameterizing the hash algorithm used is a decent idea in general, again as a long term enhancement.

**Comment by bk2204** [comment]
> > I should also point out that in many contexts the use of certain hash algorithms is requires (e.g., by government or industry standards).
> 
> That would seem to suggest parameterizing the hash algorithm used is a decent idea in general, again as a long term enhancement.

Sure, I agree that this is a logical change.  You're making this argument to the guy who has spent several years getting Git to the point where it works with SHA-256, completely in his personal time, so I think we're in violent agreement here.

My point was, more specifically, that SHA-256 is approved for most governmental uses and BLAKE2b is not.  I will admit that I'm not especially sympathetic to that situation, since there's a lot of good cryptography that is necessarily excluded for no good reason, but it is a problem we have to deal with.

In any event, I'll try to get an extension to the batch API that specifies the hash algorithm in use for 3.0, which I think is the first step in getting this to be a possibility.  So I'll mark this for the 3.0 milestone for that purpose, but that just means I'll be working on this first part, and not the entire thing.

**Comment by znull** [comment]
Is it necessary that the hash algorithm be cryptographically strong? I saw a project gain significant speedup recently by switching to a version of rsync with [xxhash](https://cyan4973.github.io/xxHash/) support.

**Comment by bk2204** [comment]
Yes, it's required that the hash be cryptographically secure.  We identify objects uniquely by a combination of hash and size (and, in some implementations, hash only), and any algorithm where collisions are feasible will result in repository corruption, since the wrong data may be smudged into the repositories otherwise.

**Comment by axkibe** [comment]
Just a note since I happened to read deeper into this once, AFAIK xxhash (or murmur hash) collisions are not feasible in the sense of more likely, contrary to cryptographically strong ones they are fakeable. (that is, it is quite possible to create/fake a binary that has a specific hash, where on the cryptographic hashing family this is a massive/almost impossible computational task.

Thus as long a use case has not to defend on people maliciously creating fake binaries with the same hash non-cryptographic algorithms should be preferred. (In my case it was interning memory objects in which case I settled for murmur since it was much simpler than xx)

**Comment by zooko** [comment]
Just a note that while SHA-256 does indeed have hardware support in modern CPUs, BLAKE3 still beats SHA-256 by a large margin in most use cases. In fact, the fact that modern CPUs have added improved SIMD hardware (like AVX512) has accelerated BLAKE3 *more* than adding SHA-256-specific instructions has accelerated SHA-256! At least in some cases. As always, benchmark your specific use case to find out what improvement would result.
